mptcp: Validating sockopt(TCP_INQ) and sockopt(IP_TOS)

spoorva opened the pull request.

dcaratti commented:
@spoorva may I ask you to rebase on top of the latesty multipath-tcp/packetdrill code? (not sure why other commits than 
`Test added for setsockopt` are present in the series). But rebasing that commit on top of the latest tree, and force-pushing again to `mptcp-net-next-tests_v2`, shoud fix the issue.
Thanks!

davide

matttbe commented on gtests/net/mptcp/unsupported/test_client_1.pkt:
```diff
@@ -0,0 +1,30 @@
+
+
+//testing mptcp for IP_TOS 
++0.0     socket(..., SOCK_STREAM, IPPROTO_MPTCP) = 7
```
detail: `socket` is not aligned with the other ones (there is one extra space before `socket`)

matttbe commented on gtests/net/mptcp/unsupported/test_client_1.pkt:
```diff
@@ -0,0 +1,30 @@
+`../../common/defaults.sh`
+
+
+//test 1 : Unsupported options
```
detail: can you add one space after `//` like in the other .pkt files please?

matttbe commented on gtests/net/mptcp/unsupported/test_client_1.pkt:
```diff
@@ -0,0 +1,30 @@
+// connect() function, connection initiated by the kernel
+--tolerance_usecs=100000
+`../../common/defaults.sh`
```
should be only one parent dir away (`../`, not `../../`), no?

How do you launch your tests? With `run_all.py`? Or with `packetdrill` binary directly. You should not launch them from the same dir. With `run_all.py`, you need to be in `gtests/net` I think. While when using `packetdrill` directly, you need to be in `gtests/net/mptcp/<dir>`

spoorva replied:
@matttbe I usually use the packetdrill directly. The resaon for " ../../" not "../" is that I put the tests in a folder in gtests/net/mptcp/unsupported where "unsupported" is the <dir> where all the tests go. I had to make the changes `../../common/defaults.sh` because the script wouldn't run otherwise.
If this is not okay, kindly let me know and I'll move the tests one folder back.

matttbe replied:
> @matttbe I usually use the packetdrill directly. The resaon for " ../../" not "../" is that I put the tests in a folder in gtests/net/mptcp/unsupported where "unsupported" is the
> where all the tests go. I had to make the changes `../../common/defaults.sh` because the script wouldn't run otherwise.
> If this is not okay, kindly let me know and I'll move the tests one folder back.

oh so you mean your test was in `gtests/net/mptcp/unsupported/<EXTRA_FOLDER>/test_client_1.pkt`? In this case, yes, you need an extra `../` to reach `gtests/net/mptcp/common` dir.
But in this PR, where it is currently located, you don't need the extra `../`

    cd gtests/net/mptcp/unsupported
    ../../packetdrill/packetdrill -v test_client_1.pkt

matttbe commented on gtests/net/mptcp/unsupported/test_client_1.pkt:
```diff
@@ -0,0 +1,30 @@
+--tolerance_usecs=100000
+`../../common/defaults.sh`
+
+
```
(detail: we don't usually have 2 blank lines in our .pkt files, no?)

matttbe commented on gtests/net/mptcp/unsupported/test_client_1.pkt:
```diff
@@ -0,0 +1,30 @@
+//test 1 : Unsupported options
+0.0     socket(..., SOCK_STREAM, IPPROTO_MPTCP) = 3
++0.0    setsockopt(3, SOL_TCP, TCP_INQ, [1], 4) = -1 EOPNOTSUPP (Operation not supported)
++0.0    getsockopt(3, SOL_TCP, TCP_INQ, [0], [4]) = -1 EOPNOTSUPP (Operation not supported)
```
I guess the code above shows what you have with the current version and the rest below is what we should have, right?

Maybe better to split: this above can go in `gtests/net/mptcp/sockopts` and what is below can stay in `unsupported`. Later, when the kernel has been modified, we can replace the one in `sockopts`.
WDYT?

About `unsupported`, maybe better to create a dedicated dir elsewhere to let people using `run_all` with the whole `mptcp` dir? e.g. in `gtests/net/mptcp_unsupported`.

spoorva replied:
@matttbe that makes sense. will do.

matttbe commented on gtests/net/mptcp/unsupported/test_client_1.pkt:
```diff
@@ -0,0 +1,30 @@
++0.0      <             S.      0:0(0)        ack 1     win 65535      <mss 1460, sackOK, TS val 700 ecr 100, nop, wscale 8, mpcapable v1 flags[flag_h] key[skey=2]>
++0.0      >  [tos 0x04] .       1:1(0)        ack 1     win 256        <nop,nop,TS val 3081837704 ecr 700, mpcapable v1 flags[flag_h] key[ckey, skey]>
+
+//+0.0  fcntl(7, F_SETFL, O_RDWR) = 0   // set back to blocking
```
why is it commented?

matttbe commented on gtests/net/mptcp/unsupported/test_client_1.pkt:
```diff
@@ -0,0 +1,30 @@
+
+//+0.0  fcntl(7, F_SETFL, O_RDWR) = 0   // set back to blocking
+
+
```
(detail: we can remove the blank lines at the end)

matttbe commented on gtests/net/mptcp/unsupported/test_client_1.pkt:
```diff
@@ -0,0 +1,30 @@
+
++0.0      >  [tos 0x04] S       0:0(0)                                 <mss 1460, sackOK, TS val 100 ecr 0,   nop, wscale 8, mpcapable v1 flags[flag_h] nokey>
++0.0      <             S.      0:0(0)        ack 1     win 65535      <mss 1460, sackOK, TS val 700 ecr 100, nop, wscale 8, mpcapable v1 flags[flag_h] key[skey=2]>
++0.0      >  [tos 0x04] .       1:1(0)        ack 1     win 256        <nop,nop,TS val 3081837704 ecr 700, mpcapable v1 flags[flag_h] key[ckey, skey]>
```
can you align the TCP options with the previous ones to make it clearer to track the modifications reading them (like in the other files) please?

Also the ACK (`.`) is no aligned with the other ACKs above

spoorva replied:
does this look right? @matttbe 

```
+0.0  >  [tos 0x04]  S   0:0(0)                    <mss 1460, sackOK, TS val 100        ecr 0,   nop, wscale 8, mpcapable v1 flags[flag_h] nokey>
+0.0  <              S.  0:0(0)  ack 1  win 65535  <mss 1460, sackOK, TS val 700        ecr 100, nop, wscale 8, mpcapable v1 flags[flag_h] key[skey=2]>
+0.0  >  [tos 0x04]   .  1:1(0)  ack 1  win 256    <nop,      nop,    TS val 3081837704 ecr 700,                mpcapable v1 flags[flag_h] key[ckey, skey]>

```

matttbe replied:
Oops, I forgot to reply: yes, perfect like that!

With a wide terminal (not limited to 80 chars), it is clearer to read and track the modifications between lines with this alignment in columns.

matttbe commented on gtests/net/mptcp/unsupported/test_client_1.pkt:
```diff
@@ -0,0 +1,30 @@
++0.0    connect(7, ..., ...) = -1 EINPROGRESS (Operation now in progress)
+
++0.0      >  [tos 0x04] S       0:0(0)                                 <mss 1460, sackOK, TS val 100 ecr 0,   nop, wscale 8, mpcapable v1 flags[flag_h] nokey>
++0.0      <             S.      0:0(0)        ack 1     win 65535      <mss 1460, sackOK, TS val 700 ecr 100, nop, wscale 8, mpcapable v1 flags[flag_h] key[skey=2]>
```
regarding the code style, we try to display everything in columns separated by two spaces (see the other files for more examples and the README file).
So here, it would be good to have 2 spaces before the `S`. Then for the seq, ack and win, you can only keep a minimum of 2 spaces between and keep them aligned.

    +0.0      >  [tos 0x04]  S   0:0(0)                    <mss (...)
    +0.0      <              S.  0:0(0)  ack 1  win 65535  <mss (...)
    (...)

matttbe commented on gtests/net/mptcp/unsupported/test_client_1.pkt:
```diff
@@ -0,0 +1,30 @@
+// connect() function, connection initiated by the kernel
```
(I'm not sure to what this comment refers to: maybe not needed?)

matttbe commented on gtests/net/mptcp/unsupported/test_client_1.pkt:
```diff
@@ -0,0 +1,30 @@
+// connect() function, connection initiated by the kernel
+--tolerance_usecs=100000
```
regarding the file name, certainly better to name it something like `ip_tos_client.pkt`

matttbe commented on gtests/net/mptcp/unsupported/test_server_1.pkt:
```diff
@@ -0,0 +1,15 @@
+
+// connection initiated by packetdrill
```
(probably not needed)

matttbe commented on gtests/net/mptcp/unsupported/test_server_1.pkt:
```diff
@@ -0,0 +1,15 @@
+
```
same here for the filename

matttbe commented on gtests/net/mptcp/unsupported/test_server_1.pkt:
```diff
@@ -0,0 +1,15 @@
++0     listen(3, 1) = 0
++0	 <  S   0:0(0)                     win 32792  <mss 1460, sackOK, nop, nop, nop, wscale 7, mpcapable v1 flags[flag_h] nokey>
++0	 >  S.  0:0(0)           ack 1                <mss 1460, nop, nop, sackOK, nop, wscale 8, mpcapable v1 flags[flag_h] key[skey]>
++0.01    <   .  1:1(0)           ack 1     win 257                                               <mpcapable v1 flags[flag_h] key[ckey=2, skey]>
```
please fix the alignment with the 2 prev lines (using spaces, no tabs)

matttbe commented on gtests/net/mptcp/unsupported/test_server_1.pkt:
```diff
@@ -0,0 +1,15 @@
+
+
++0     socket(..., SOCK_STREAM, IPPROTO_MPTCP) = 3
++0     setsockopt(3, SOL_SOCKET, SO_ZEROCOPY, [1], 4) = 0
```
I guess this is unrelated to IP_TOS and this file should not be in this commit, right?

matttbe commented on gtests/net/mptcp/unsupported/test_server_1.pkt:
```diff
@@ -0,0 +1,15 @@
+
+// connection initiated by packetdrill
+--tolerance_usecs=100000
+`../../common/defaults.sh`
```
same here for `../../`

matttbe commented on gtests/net/mptcp/unsupported/test_server_1.pkt:
```diff
@@ -0,0 +1,15 @@
+// connection initiated by packetdrill
+--tolerance_usecs=100000
+`../../common/defaults.sh`
+
+
```
same here: typically we don't have double blank lines

matttbe commented on gtests/net/mptcp/unsupported/test_server_1.pkt:
```diff
@@ -0,0 +1,15 @@
+
++0     bind(3, ..., ...) = 0
++0     listen(3, 1) = 0
++0	 <  S   0:0(0)                     win 32792  <mss 1460, sackOK, nop, nop, nop, wscale 7, mpcapable v1 flags[flag_h] nokey>
```
please fix the alignment like in the other files: one blank line before this one + 2 spaces before `<` to have an indentation compared to the syscalls.

matttbe reviewed (changes requested):
Thank you for looking at that!

I just did a quick review and put everything that passed to my head. Do not worry there are a bunch of comments, it is normal for a first PR here, especially with packetdrill. I hope my comments will help. Do not hesitate to reply if something is not clear.

For the "code style", do not hesitate to check the README file and get inspiration from other files.

One last thing: the commit message is very important. It should give very quick indication in <50 chars for the title, explain what you did but more importantly why you did that. Understanding what you changed is easy (but good to see if this is aligned with the code) but the reason is often harder to guess.
Any other valuable info should be there too, e.g. currently the kernel doesn't support that → hence the 'unsupported' file, what we should see, if you are based on another pkt file, etc.
Also please do not forget your Signed-off-by tag ;)

matttbe commented on gtests/net/mptcp/sockopts/test_unsupported_sockopts_tcp_inq.pkt:
```diff
@@ -0,0 +1,7 @@
+-tolerance_usecs=100000
```
regarding the name of the new files, they look good to me but I see the others we have don't have this `test_` prefix. May you remove them please?

e.g.

    test_mptcp_ip_tos_client.pkt → mptcp_ip_tos_client.pkt

spoorva replied:
got it.

matttbe replied:
(I had a wrong example in my previous comment, fixed now but I'm sure you understood what I wanted to say ;) )

matttbe commented on gtests/net/mptcp/sockopts/test_unsupported_sockopts_tcp_inq.pkt:
```diff
@@ -0,0 +1,7 @@
+-tolerance_usecs=100000
+`../../common/defaults.sh`
+
+// test 1 : Unsupported options
```
I don't think this comment brings any value. Maybe better with something like:

```c
// TCP_INQ is not supported with MPTCP sockets yet.
// See <github issue URL> to track the evolution.
// Once supported, we can replace this test by the one in:
//   gtests/net/mptcp_unsupported/mptcp_tcp_inq_client.pkt
```

spoorva replied:
@matttbe I can't find any open issue for TCP_INQ but I checked if TCP_INQ is implemented yet - it's not.
https://github.com/multipath-tcp/mptcp_net-next/blob/079dbabc3d57af1a8bce79cd107b2291aa8b9dcc/net/mptcp/sockopt.c#L551

matttbe replied:
We can open a new one then ;-)
https://github.com/multipath-tcp/mptcp_net-next/issues/224

Always best to have an issue to track and regroup comments around this feature.

matttbe commented on gtests/net/mptcp/sockopts/test_unsupported_sockopts_tcp_inq.pkt:
```diff
@@ -0,0 +1,7 @@
+// test 1 : Unsupported options
+0.0   socket(..., SOCK_STREAM, IPPROTO_MPTCP) = 3
++0.0  setsockopt(3, SOL_TCP, TCP_INQ, [1], 4) = -1 EOPNOTSUPP (Operation not supported)
++0.0  getsockopt(3, SOL_TCP, TCP_INQ, [0], [4]) = -1 EOPNOTSUPP (Operation not supported)
```
Maybe best to check in the commit only IP_TOS and do TCP_INQ in another one, no?

spoorva replied:
I am sorry, I am not sure I understand. Do we include the TCP_INQ one in another commit?

matttbe replied:
My bad, I wanted to say: `s/check/keep`.

In your commit message, you mention you add support for `IP_TOS`  but here this file is not about `IP_TOS` but `TCP_INQ`. Best to have one commit dedicated to `IP_TOS`. If you want to support `TCP_INQ`, please do it in a separated commit.

spoorva replied:
Ah, I understand. Maybe I should edit my commit message to say that there are a bunch of tests and not just to check IP_TOS. Or would it be better to have the TCP_INQ testin another commit?

spoorva replied:
After this is resolved, I'll squash the commits and push the changes all together.

matttbe replied:
> Or would it be better to have the TCP_INQ testin another commit?

One commit per "theme" is usually preferred.

So ideally, each commit would have 2 tests:
- one in `mptcp/sockopts` validating the current situation for one sockopt: `EOPNOTSUPP`
- one in `mptcp_unsupported` validation what we should have if this sockopt is supported with MPTCP sockets.

spoorva replied:
@mattbe I am not entirely sure what to do. I have an idea -
1. I put one test (let's say the one in mptcp/sockopts) in a new commit and squash all my previous commits together with this one. But in this case, I would still have the old files(mptcp_unsupported) present I think? And then do another commit for mptcp_unsupported.

OR

2. I could do a git reset to the upstream head and start the process all over.

wdyt would be a better idea?

matttbe replied:
Up to you but from what I see on this PR, it seems needed to rewrite the commits. If you prefer to "reset and commit" or "rebase", that's the same. If you are not used to manipulate Git, it might look easier to use:

```
git reset --soft origin/mptcp-net-next
git restore --staged .
git add -p
git commit -s
```

But you can always try to do: `git rebase -i origin/mptcp-net-next` and follow the instructions. Maybe best to create a "backup" branch (`git branch backup`) before doing that so you can easily find the previous version you had before (even if `git status` will also tell you to use `git rebase --abort` to go back to the previous state).

matttbe commented on gtests/net/mptcp_unsupported/test_mptcp_ip_tos_client.pkt:
```diff
@@ -0,0 +1,21 @@
++0.0  <              S.  0:0(0)  ack 1  win 65535  <mss 1460, sackOK, TS val 700        ecr 100, nop, wscale 8, mpcapable v1 flags[flag_h] key[skey=2]>
++0.0  >  [tos 0x04]   .  1:1(0)  ack 1  win 256    <nop,      nop,    TS val 3081837704 ecr 700,                mpcapable v1 flags[flag_h] key[ckey, skey]>
+
++0.0  fcntl(7, F_SETFL, O_RDWR) = 0   // set back to blocking
```
(detail: maybe not needed?)

matttbe commented on gtests/net/mptcp_unsupported/test_mptcp_ip_tos_client.pkt:
```diff
@@ -0,0 +1,21 @@
++0.0  >  [tos 0x04]   .  1:1(0)  ack 1  win 256    <nop,      nop,    TS val 3081837704 ecr 700,                mpcapable v1 flags[flag_h] key[ckey, skey]>
+
++0.0  fcntl(7, F_SETFL, O_RDWR) = 0   // set back to blocking
+
```
please remove these 2 empty lines at the end

matttbe commented on gtests/net/mptcp_unsupported/test_mptcp_ip_tos_server.pkt:
```diff
@@ -0,0 +1,14 @@
+`../common/defaults.sh`
+
++0  socket(..., SOCK_STREAM, IPPROTO_MPTCP) = 3
++0  setsockopt(3, SOL_SOCKET, SO_ZEROCOPY, [1], 4) = 0
```
Why do you need zerocopy?

Do you not need to set `IP_TOS` instead?

spoorva replied:
ah, sorry, my bad.

matttbe commented on gtests/net/mptcp_unsupported/test_mptcp_ip_tos_server.pkt:
```diff
@@ -0,0 +1,14 @@
+
++0     <  S   0:0(0)         win 32792  <mss 1460, sackOK, nop, nop,    nop, wscale 7, mpcapable v1 flags[flag_h] nokey>
++0     >  S.  0:0(0)  ack 1             <mss 1460, nop,    nop, sackOK, nop, wscale 8, mpcapable v1 flags[flag_h] key[skey]>
++0.01  <   .  1:1(0)  ack 1  win 257                                                  <mpcapable v1 flags[flag_h] key[ckey=2, skey]>
```
it looks like you don't check anything related to `tos` here

spoorva replied:
fixed.

matttbe reviewed (changes requested):
Thank you for the update.

Do you mind replying on the previous comments on Github? e.g. if you applied the modifications that were asked (Resolve conversation) or reply with a comment if something is not clear, etc?
When the review cycles are not done the same day, it is not easy to remember the previous comments :-)
On Github, we have this nice feature allowing us to mark comments as "Done", better to use it ;-)

Now regarding the commits, in the kernel dev -- and linked tools -- we use to amend/rewrite the commits if modifications are needed and spot during the review. In other words, if you need to fix something introduced by your in-review patch, best to modify this patch directly and update the commit message if needed. But then no need to create a new commit fixing issues of the previous commit if this previous commit has not been merged yet.
In you case here, you can then squash the two commits: `git rebase -i origin/mptcp-net-next` → squash → rewrite commit message if needed.

(just in case you need more details about this workflow, do not hesitate to ask questions here or on IRC or https://www.atlassian.com/git/tutorials/rewriting-history or others ; I know it can be quite disturbing at the beginning and I don't know if you did that a lot before)

matttbe commented:
Thank you for your replies!

You mentioned "fixed" but I don't see any modification. No hurry, just to know if you didn't miss any `git push -f` ;-)

One last thing: for the commit message, please try to follow the "standards" by keeping the title under ~50 chars and the rest of the message under ~72 chars per line.
You can get some examples by looking at the history. Something like `mptcp: validating sockopt(IP_TOS)` seems like a good short summary for a title. In the description, do not forget to add the reason why your patch is needed: the only info that is not always easy to get by looking at the code. It doesn't need to be long if it is obvious, e.g.

    This sockopt was not validated before.
    It is currently not supported by MPTCP but its implementation on the kernel side is in progress, see issue <GH URL>.

(We cannot "easily" get this info just by looking at the code so that's a very info to add in the commit message for the reviewers and devs, including you later ;) )

matttbe commented on gtests/net/mptcp/mp_capable/v1_mp_capable_bind_no_cs_ooo.pkt:
```diff
@@ -0,0 +1,14 @@
++0       <  S   0:0(0)         win 32792  <mss 1000, sackOK, nop, nop, nop, wscale 7, mpcapable v1 flags[flag_h] nokey>
++0       >  S.  0:0(0)  ack 1             <mss 1460, nop, nop, sackOK, nop, wscale 8, mpcapable v1 flags[flag_h] key[skey]>
++0.01    <   .  101:201(100)   ack 1  win 257       <mpcapable v1 flags[flag_h] key[ckey=2, skey] mpcdatalen 200, nop, nop>
++0       >   .  1:1(0)         ack 1      <nop, nop, sack 101:201, dss dack4=1 nocs>
```
Should we align the `ack`, `win` and TCP Options like in the other files?

```
+0       <  S   0:0(0)               win 32792  <mss 1000, sackOK, nop, nop, nop, wscale 7, mpcapable v1 flags[flag_h] nokey>
+0       >  S.  0:0(0)        ack 1             <mss 1460, nop, nop, sackOK, nop, wscale 8, mpcapable v1 flags[flag_h] key[skey]>
+0.01    <   .  101:201(100)  ack 1  win 257                                               <mpcapable v1 flags[flag_h] key[ckey=2, skey] mpcdatalen 200, nop, nop>
+0       >   .  1:1(0)        ack 1             <nop, nop, sack 101:201, dss dack4=1 nocs>
```

matttbe replied:
EDIT: You can ignore my previous comment ↑
I'm not sure what you did exactly but this file is already in `mptcp-net-next` branch.
It looks like you modified the commit because the original sha is 067597d46f92c74e2ccaf4d258fbb3011ef48fe7

May you rebase your branch on top of latest `origin/mptcp-net-next` branch please?
(or re-start from a new branch and cherry-pick only your own commits on top)

I just opened a new PR fixing the code style.

spoorva replied:
@matttbe I apologize for the confusion. I mam not entirely surely what I did but it all looks weird. Fixing it asap.

matttbe replied:
All good, it happens ;-)

matttbe commented on gtests/net/mptcp/common/multi-ep.sh:
```diff
@@ -65,7 +65,7 @@ else # ipv4 or ipv4-mapped-ipv6
     if [[ $OPT_LOCAL_IP =~ ([0-9]+[.][0-9]+[.][0-9]+[.])([0-9]+) ]]; then
```
detail about the commit title (`fix name of netmask environment variable`), may you prefix it `mptcp` please?
To clearly identify which commits are specific to us and not from upstream.

matttbe replied:
EDIT: You can ignore my previous comment ↑
This commit has already been merged. The previous comment is then just for @dcaratti for next time ;-)

matttbe commented on gtests/net/mptcp_unsupported/mptcp_ip_tos_client.pkt:
```diff
@@ -0,0 +1,17 @@
+--tolerance_usecs=100000
```
detail about the commit title: please prefix it with `mptcp:`, it is easier to find commits that are not from upstream.

    mptcp: validating sockopt(IP_TOS)

matttbe commented on gtests/net/mptcp_unsupported/mptcp_ip_tos_server.pkt:
```diff
@@ -0,0 +1,14 @@
+`../common/defaults.sh`
+
++0  socket(..., SOCK_STREAM, IPPROTO_MPTCP) = 3
++0  setsockopt(3, SOL_IP, IP_TOS, [4], 4) = 0
```
maybe best to also check the expected returned value of a `getsockopt(IP_TOS)` like you did in the client side.

matttbe commented on gtests/net/mptcp_unsupported/mptcp_ip_tos_server.pkt:
```diff
@@ -0,0 +1,14 @@
++0  bind(3, ..., ...) = 0
++0  listen(3, 1) = 0
+
++0.0  >  [tos 0x04]  S   0:0(0)                    <mss 1460, sackOK, TS val 100        ecr 0,   nop, wscale 8, mpcapable v1 flags[flag_h] nokey>
```
this is the output of a `connect()` (client side) but not a `bind()/listen()` (server side).

matttbe commented on gtests/net/mptcp_unsupported/mptcp_ip_tos_server.pkt:
```diff
@@ -0,0 +1,14 @@
+--tolerance_usecs=100000
```
@dcaratti : should we have a version in `mptcp/sockopts` showing the current status? I mean: showing that the `[sg]etsockopt()` are showing that the sockopt is not supported so it would fail when it will implemented? (so we don't forget to use this one from the `unsupported` dir?

It is optional, maybe we will not forget about the packetdrill test when implementing this sockopt :)

dcaratti replied:
> @dcaratti : should we have a version in `mptcp/sockopts` showing the current status? I mean: showing that the `[sg]etsockopt()` are showing that the sockopt is not supported so it would fail when it will implemented? (so we don't forget to use this one from the `unsupported` dir?

not sure if it's worth doing. I initially thought to test "unsupported" sockopt only for those that are unsupported by design on MPTCP (and, I thought `TCP_INQ` was a good example :-) ).

matttbe replied:
Yes sure, we can only have one and remember about adding new packetdrill tests when supporting more sockopts :)

matttbe commented on gtests/net/mptcp/sockopts/mptcp_unsupported_sockopts_tcp_inq.pkt:
```diff
@@ -0,0 +1,9 @@
+-tolerance_usecs=100000
```
may you also prefix your commit with `mptcp:` please?

matttbe commented on gtests/net/mptcp/sockopts/mptcp_unsupported_sockopts_tcp_inq.pkt:
```diff
@@ -0,0 +1,9 @@
+
+// test 1 : TCP_INQ is not supported by MPTCP sockets yet.
+// Follow https://github.com/multipath-tcp/mptcp_net-next/issues/224 for progress 
+// Once supported, we can replace it with gtests/net/mptcp_unsupported/mptcp_tcp_inq_client.pkt 
```
this file doesn't exist. Did you forget to add it to git?

matttbe replied:
> this file doesn't exist. Did you forget to add it to git?

Is the file missing or should we remove this line?

spoorva replied:
@matttbe ah yes, I think we need to remove this for now because I don't have a test in  gtests/net/mptcp_unsupported/ to check for TCP_INQ when it's supported. Do you think it's a good idea to add the test or remove the line in question?

matttbe replied:
The best would be to add a test but for now -- and if the main goal is to check EOPNOTSUPP is returned for unsupported sockopts -- we can simply remove this line.

spoorva replied:
I guess the main goal was to check for EOPNOTSUPP. I'll go ahead and remove the line.

matttbe reviewed (changes requested):
Thank you for the update!

It looks like you added 4 extra commits that were already in the repo. May you do a rebase please?

dcaratti commented on gtests/net/mptcp_unsupported/mptcp_ip_tos_client.pkt:
```diff
@@ -0,0 +1,17 @@
+--tolerance_usecs=100000
+`../common/defaults.sh`
+
```
probably it's not strictly needed on the 3WHS, but it should be safer to add --tolerance_usecs=<big value> (so that the test doesn't fail on slow machines)

dcaratti replied:
oh wait. scratch this comment

dcaratti commented on gtests/net/mptcp_unsupported/mptcp_ip_tos_server.pkt:
```diff
@@ -0,0 +1,15 @@
++0  listen(3, 1) = 0
+
++0.0  <  [tos 0x04]  S   0:0(0)                    <mss 1460, sackOK, TS val 100        ecr 0,   nop, wscale 8, mpcapable v1 flags[flag_h] nokey>
++0.0  >              S.  0:0(0)  ack 1  win 65535  <mss 1460, sackOK, TS val 700        ecr 100, nop, wscale 8, mpcapable v1 flags[flag_h] key[skey=2]>
```
please note, the [tos 0x4] is validated on outbound packet: unless you add it, this test will pass also on unpatched kernels (because the default tos is 0x0)

matttbe replied:
so we should have the opposite I guess, something like this:

```
+0.0  <              S   0:0(0)                    <mss 1460, sackOK, TS val 100        ecr 0,   nop, wscale 8, mpcapable v1 flags[flag_h] nokey>
+0.0  >  [tos 0x04]  S.  0:0(0)  ack 1  win 65535  <mss 1460, sackOK, TS val 700        ecr 100, nop, wscale 8, mpcapable v1 flags[flag_h] key[skey=2]>
+0.0  <               .  1:1(0)  ack 1  win 256    <nop,      nop,    TS val 3081837704 ecr 700,                mpcapable v1 flags[flag_h] key[ckey, skey]>

spoorva replied:
ah got it. We need to set it in an outbound packet. Thanks @dcaratti @matttbe

matttbe reviewed (changes requested):
Hi Poorva,

Thank you for the update!

There is one opened comment left, a small thing, it will be quick to fix, please see below.

Also just two small things for next time:
- for the commit title, please prefix them with `mptcp: ` including a space after `:`, e.g. `mptcp: foo bar` and not `mptcp:foo bar`
- please mark comments as resolved on Github (`Resolve Conversation` button) after having pushed new commits fixing what was discussed in these comments. This is really helping the reviewers, especially when the cycles are long like here due to the timezone difference ;-)

spoorva commented:
> Hi Poorva,
> 
> Thank you for the update!
> 
> There is one opened comment left, a small thing, it will be quick to fix, please see below.
> 
> Also just two small things for next time:
> 
>     * for the commit title, please prefix them with `mptcp: ` including a space after `:`, e.g. `mptcp: foo bar` and not `mptcp:foo bar`

Yes, will do.

> 
>     * please mark comments as resolved on Github (`Resolve Conversation` button) after having pushed new commits fixing what was discussed in these comments. This is really helping the reviewers, especially when the cycles are long like here due to the timezone difference ;-)

got it!

matttbe commented on gtests/net/mptcp/sockopts/mptcp_unsupported_sockopts_tcp_inq.pkt:
```diff
@@ -0,0 +1,8 @@
+`../../common/defaults.sh`
+
+// test 1 : TCP_INQ is not supported by MPTCP sockets yet.
+// Follow https://github.com/multipath-tcp/mptcp_net-next/issues/224 for progress  
```
sorry to be picky about that but please remove trailing whitespace and configure your editor to do that automatically. Most of us have this configuration. It means that if we need to edit this file, we will end up with "undesired" modifications.

(it is a shame Github doesn't clearly show them: I noticed them because you just modified this line with one more space at the end and I was wondering why you edited this line, it didn't seem modified.)

matttbe commented on gtests/net/mptcp_unsupported/mptcp_ip_tos_client.pkt:
```diff
@@ -0,0 +1,17 @@
+--tolerance_usecs=100000
+`../common/defaults.sh`
+
+// testing mptcp for IP_TOS 
```
same here for the trailing whitespace

matttbe commented on gtests/net/mptcp_unsupported/mptcp_ip_tos_server.pkt:
```diff
@@ -0,0 +1,15 @@
+
++0      socket(..., SOCK_STREAM, IPPROTO_MPTCP) = 3
++0      setsockopt(3, SOL_IP, IP_TOS, [4], 4) = 0
++0.200  getsockopt(7, SOL_IP, IP_TOS, [4], [4]) = 0 
```
and here

spoorva replied:
@matttbe done!

matttbe reviewed (changes requested):
one last thing to fix and we are good!

@dcaratti : is it also OK for you?

spoorva committed:
mptcp: Adding test for sockopt(TCP_INQ)

Currently, TCP_INQ is not supported by MPTCP.
Follow multipath-tcp/mptcp_net-next#224 for progress

Signed-off-by: Poorva Sonparote <psonparo@redhat.com>

spoorva committed:
mptcp: Validating sockopt(IP_TOS)

This sockopt was not validated before.
Currently not supported by MPTCP but it's implementation is in progress.
See issue multipath-tcp/mptcp_net-next#220

Signed-off-by: Poorva Sonparote <psonparo@redhat.com>

matttbe reviewed (approved):
Thank you for the multiple updates! It looks good to me!

@dcaratti Same for you?

dcaratti reviewed (approved):
yes, now it looks ok. Thanks!
